### ISSUE TITLE
Proposal: Expose `llama_get_logits`-based method on Context in high level API

### ISSUE THREAD

**Issue by brittlewis12** [utilityai/llama-cpp-rs#505]
Today, llama-cpp-2 exposes `llama_get_logits_ith`-based `candidates_ith`.

In working with the underlying library, I have leaned on the slightly different  `llama_get_logits`. It seems that while the underlying library accepts `-1` as input to the `-ith` variant to accomplish the same as `llama_get_logits`, the current implementation of safely ensuring logits are initialized for the given `llama_pos` disallows `-1` as input as a side effect.

Honestly, the existing implementation is sound as is, and I'd prefer to use the other, slightly simpler seam either way.

I took a quick stab at what this could look like, based on the existing pattern of a lower-level `unsafe` call and a higher level wrapper that returns an unsorted `LlamaTokenData` iter for you.
- [Expose last-token candidates context method](https://github.com/brittlewis12/llama-cpp-rs/pull/1)

I [tweaked the existing](https://github.com/brittlewis12/llama-cpp-rs/compare/candidates_last...brittlewis12:llama-cpp-rs:candidates-last-simple?expand=1#diff-7319277c084bff5c625f04b23366571e88d4463276613007bee0d70b72cbc445) `simple` example to use this method without issue:
```
-            let candidates = ctx.candidates_ith(batch.n_tokens() - 1);
+            let candidates = ctx.candidates_last();
```

Thank you as ever for your maintenance efforts here @MarcusDunn!

---

a little more background [via llama.cpp](https://github.com/ggerganov/llama.cpp/blob/c35e586ea57221844442c65a1172498c54971cb0/include/llama.h#L844-L849):

> Token logits obtained from the last call to llama_decode()
> The logits for which llama_batch.logits[i] != 0 are stored contiguously in the order they have appeared in the batch.
> Rows: number of tokens for which llama_batch.logits[i] != 0
> Cols: n_vocab

**Comment by MarcusDunn** [comment]
whoops, thought that was a PR, didn't mean to review your fork!

**Comment by brittlewis12** [comment]
no problem 😆 will open it up properly on your end

**Comment by brittlewis12** [comment]
re: https://github.com/brittlewis12/llama-cpp-rs/pull/1#pullrequestreview-2323689876
would removing “last” from the names, so `get_logits` & just `candidates`, be more in line with your preferred naming conventions?

**Comment by MarcusDunn** [comment]
yeah, that matches what I've done elsewhere a more closely.

**Comment by brittlewis12** [comment]
thanks @MarcusDunn!

I have a few more simple changes ready to go — would you prefer independent issues for each, or just PRs?

- expose `llama_kv_cache_seq_rm` context method
- expose `flash_attn` context param
- expose `offload_kqv` context param
- expose `llama_token_is_eog` model method
- expose mirostat v1 sampling

**Comment by MarcusDunn** [comment]
Feel free to make that a single PR.
